modules/aruco/src/aruco.cpp

```diff
@@ -0,0 +1,1573 @@
+                                    vector< vector< vector< Point > > > *_contoursArrays,
+                                    DetectorParameters *_params)
+        : grey(_grey), candidatesArrays(_candidatesArrays), contoursArrays(_contoursArrays),
+          params(_params) {}
+
+    void operator()(const Range &range) const {
+        const int begin = range.start;
+        const int end = range.end;
+
+        for(int i = begin; i < end; i++) {
+            int currScale =
+                params->adaptiveThreshWinSizeMin + i * params->adaptiveThreshWinSizeStep;
+            // threshold
+            Mat thresh;
+            _threshold(*grey, thresh, currScale, params->adaptiveThreshConstant);
+
+            // detect rectangles
+            _findMarkerContours(thresh, (*candidatesArrays)[i], (*contoursArrays)[i],
+                                params->minMarkerPerimeterRate, params->maxMarkerPerimeterRate,
+                                params->polygonalApproxAccuracyRate, params->minCornerDistanceRate,
+                                params->minDistanceToBorder);
+        }
+    }
+
+    private:
+    DetectInitialCandidatesParallel &operator=(const DetectInitialCandidatesParallel &);
+
+    const Mat *grey;
+    vector< vector< vector< Point2f > > > *candidatesArrays;
+    vector< vector< vector< Point > > > *contoursArrays;
+    DetectorParameters *params;
+};
+
+
+/**
+ * @brief Initial steps on finding square candidates
+ */
+static void _detectInitialCandidates(const Mat &grey, vector< vector< Point2f > > &candidates,
+                                     vector< vector< Point > > &contours,
+                                     DetectorParameters params) {
+
+    CV_Assert(params.adaptiveThreshWinSizeMin >= 3 && params.adaptiveThreshWinSizeMax >= 3);
+    CV_Assert(params.adaptiveThreshWinSizeMax >= params.adaptiveThreshWinSizeMin);
+    CV_Assert(params.adaptiveThreshWinSizeStep > 0);
+
+    // number of window sizes (scales) to apply adaptive thresholding
+    int nScales = (params.adaptiveThreshWinSizeMax - params.adaptiveThreshWinSizeMin) /
+                      params.adaptiveThreshWinSizeStep + 1;
+
+    // if only one scale
+    if(nScales == 1) {
+        int scale = params.adaptiveThreshWinSizeMin;
+        // treshold
+        Mat thresh;
+        _threshold(grey, thresh, scale, params.adaptiveThreshConstant);
+
+        // detect rectangles
+        vector< vector< Point2f > > currentCandidates;
+        vector< vector< Point > > currentContours;
+        _findMarkerContours(thresh, currentCandidates, currentContours,
+                            params.minMarkerPerimeterRate, params.maxMarkerPerimeterRate,
+                            params.polygonalApproxAccuracyRate, params.minCornerDistanceRate,
+                            params.minDistanceToBorder);
+
+        // join candidates
+        for(unsigned int i = 0; i < currentCandidates.size(); i++) {
+            candidates.push_back(currentCandidates[i]);
```

Comment:
Why do you have a join candidates in the single scale case? You could just pass a reference.